datalad/distribution/publish.py

```diff
@@ -127,63 +133,93 @@ def __call__(dataset=None, dest=None, path=None,
                              "{0}.".format(ds.path))
         assert(ds.repo is not None)
 
-        # TODO: For now we can deal with a sibling(remote) name given by `dest`
-        # only. Figure out, when to allow for passing a local path or URL
-        # directly and what to do in that case.
-
-        # Note: we need an upstream remote, if there's none given. We could
-        # wait for git push to complain, but we need to explicitly figure it
-        # out for pushing annex branch anyway and we might as well fail right
-        # here.
-
-        # keep original dest in case it's None for passing to recursive calls:
-        dest_resolved = dest
-        if dest is None:
-            # check for tracking branch's remote:
-            try:
-                std_out, std_err = \
-                    ds.repo._git_custom_command('',
-                                                ["git", "config", "--get", "branch.{active_branch}.remote".format(active_branch=ds.repo.get_active_branch())],
-                                                expect_fail=True)
-            except CommandError as e:
-                if e.code == 1 and e.stdout == "":
-                    std_out = None
+        # figure out, what to publish from what (sub)dataset:
+        publish_this = False   # whether to publish `ds`
+        publish_files = []     # which files to publish by `ds`
+
+        expl_subs = set()      # subdatasets to publish explicitly
+        publish_subs = dict()  # collect what to publish from subdatasets
+        if recursive:
+            for subds_path in ds.get_subdatasets(fulfilled=True):
+                if path and '.' in path:
+                    # we explicitly are passing '.' to subdatasets in case of
+                    # `recursive`. Therefore these datasets are going into
+                    # `publish_subs`, instead of `expl_subs`:
+                    sub = Dataset(opj(ds.path, subds_path))
+                    publish_subs[sub.path] = dict()
+                    publish_subs[sub.path]['dataset'] = sub
+                    publish_subs[sub.path]['files'] = ['.']
```

Comment:
I will merge, but I wondered why 3 lines of code with duplications instead of e.g. 
```python
publish_subs[sub.path] = {'dataset': sub, 'files': ['.']}
```